File: 2563.count-the-number-of-fair-pairs.py
#
# @lc app=leetcode id=2563 lang=python3
#
# [2563] Count the Number of Fair Pairs
#

# @lc code=start
import bisect
class Solution:
    def countFairPairs(self, nums: list[int], lower: int, upper: int) -> int:
        ans = 0
        nums.sort()
        for i in range(len(nums)):
            r = bisect.bisect_right(nums, upper - nums[i], 0, i)
            l = bisect.bisect_left(nums, lower - nums[i], 0, i)
            ans += r - l 
        return ans
# @lc code=end

File: 2563.count-the-number-of-fair-pairs.java
/*
 * @lc app=leetcode id=2563 lang=java
 *
 * [2563] Count the Number of Fair Pairs
 */
import java.util.Arrays;
// @lc code=start
class Solution {
    public long countFairPairs(int[] nums, int lower, int upper) {
        //可以使用枚举策略 
        //两次想向双指针优化策略
        Arrays.sort(nums);
        return count(nums, upper) - count(nums, lower - 1);

    }
    private long count(int[] nums, int upper){
        long ans = 0; 
        int n = nums.length;
        int right = n -1;
        for(int i = 0; i < n; i++){
            while(right > i && nums[right] > upper - nums[i]){
                right--;
            }
            if(right == i) break;
            ans += right - i;
        }
        return ans;
    }
}
// @lc code=end

